pkg/skaffold/sync/sync.go

```diff
@@ -86,11 +99,56 @@ func NewItem(a *latest.Artifact, e filemon.Events, builds []build.Artifact, inse
 		return nil, nil
 	}
 
-	return &Item{
-		Image:  tag,
-		Copy:   toCopy,
-		Delete: toDelete,
-	}, nil
+	return &Item{Image: tag, Copy: toCopy, Delete: toDelete}, nil
+}
+
+func inferredSyncItem(a *latest.Artifact, e filemon.Events, builds []build.Artifact, provider DestinationProvider) (*Item, error) {
+	// deleted files are no longer contained in the syncMap, so we need to rebuild
+	if len(e.Deleted) > 0 {
+		return nil, nil
+	}
+
+	tag := latestTag(a.ImageName, builds)
+	if tag == "" {
+		return nil, fmt.Errorf("could not find latest tag for image %s in builds: %v", a.ImageName, builds)
+	}
+
+	syncMap, err := provider()
+	if err != nil {
+		return nil, errors.Wrapf(err, "infer syncmap for image %s", a.ImageName)
+	}
+
+	toCopy := make(map[string][]string)
+	for _, f := range append(e.Modified, e.Added...) {
+		relPath, err := filepath.Rel(a.Workspace, f)
+		if err != nil {
+			return nil, errors.Wrapf(err, "changed file %s can't be found relative to context %s", f, a.Workspace)
```

Comment:
nit: `errors.Wrapf(err, "finding changed file %s relative to context %s", f, a.Workspace)`